Documentation for OQL

dahfjkg opened the pull request:
This is the documentation for OQL following the new structure.

It contains the original Clauses PR and changes from [OQL expressions](https://github.com/mendix/docs/pull/8434/commits) and [OQL expression syntax](https://github.com/mendix/docs/pull/8520/commits)

pijuskri committed:
Combine existing content into oql-expressions doc

pijuskri committed:
Parameter example improvements

pijuskri committed:
Aggregate documentation

pijuskri committed:
Aggregate documentation, examples

pijuskri committed:
Combine functions and operators into a single file

dahfjkg committed:
Documentation for OQL clauses

pijuskri committed:
Operator structure, Plus, Minus

pijuskri committed:
Binary operator examples, null handling, precedence, string coercion

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+| NULL     | -1     |
+
+
+### `WHERE` clause
```
Good to mention that `WHERE NULL` is equivalent to `WHERE FALSE`

pijuskri replied:
Can be resolved

pijuskri committed:
Correct parameter examples, remove bad examples

pijuskri committed:
Complete other binary operators

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+
+These basic clauses define which data needs to be retrieved, and which entity or entities should be used as a source. Other clauses are not mandatory. They are used to specify limitations and other rules on the data being retrieved.
+
+Other clauses are not mandatory, but if used, they should be present in a fixed order:
```
"Other clauses are not mandatory" is repeated in line 34

pijuskri replied:
Can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+
+
+{{% alert color="info" %}}
+In some scenarios, aliases in `FROM` are mandatory. Aliases are mandatory when defining a view entity.
```
"In some scenarios, aliases in `FROM` are mandatory." should probably be moved to the FROM clause section

pijuskri replied:
Can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+| -1            | 0             |
+
+{{% alert color="info" %}}
+Please note that if you use a subquery as an expression in `FROM`, then such subquery should always return  a single column, and the number of rows should be at most one. If the subquery returns more than one row or a number of columns different than one, that will lead to an exception during runtime.
```
"in FROM" should be "in SELECT"

pijuskri replied:
Can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+| 562949953421521   | John      | Doe      | 1688849860264073 | Doe          | 1      |
+| 562949953421923   | Jane      | Moose    | 1688849860264231 | Moose        | 2      |
+
+It is possible to specify columns in `SELECT`. To avoid ambiguity in case opf duplicate attribute names, it is possible to specify entity name or alias. Respectively, `<module>.<entity>/<attribute>` or `<alias>/<attribute>`. It is possible to retrieve alll attributes of a particular entity using `<alias>/*`
```
A lot of overlap with what is already mentioned in `SELECT` clause.

pijuskri replied:
This can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+	entity_path | entity_name | ( sub_oql_query ) [ [ AS ] from_alias ]
+	[ ON <constraint> ]
+
+##### entity_path
```
Should mention `ON` which is mandatory when long path is not used

pijuskri replied:
This can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
````diff
@@ -0,0 +1,998 @@
+WHERE <constraint>
+```
+
+`<constraint>` is an expression for which the value always equals true. Expressions consist of simple comparisons using operators, functions, keywords or system variables.
````
What do you mean by "for which the value is always true"? Perhaps it should have been "which is of type BOOLEAN"?

dahfjkg replied:
By this I mean the same as the author of the original docs: nothing that makes sense 😅

pijuskri replied:
This can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
````diff
@@ -0,0 +1,998 @@
+WHERE <constraint>
+```
+
+`<constraint>` is an expression for which the value always equals true. Expressions consist of simple comparisons using operators, functions, keywords or system variables.
````
Are system variables parameters? Do we call them that way?

dahfjkg replied:
System variables are a different thing https://docs.mendix.com/refguide/xpath-keywords-and-system-variables/#system-variables

pijuskri replied:
But how can they be used in OQL?

dahfjkg replied:
Like this: `select Name from Module.Person where Module.Person_USER = '[%CurrentUser%]'`

There is a test for this

pijuskri replied:
This can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+
+The `GROUP BY` clause is used to group OQL query results into summary rows based on the values of one or multiple attribuites. It is possible to filter aggregated results further using a `HAVING` clause.
+
+#### Syntax
```
Good to explain what is allowed in the expression (any columns of entities in scope) and HAVING constraint (aggregates of columns in group by). But this might not be neccesary as this behaviour is explained later.

dahfjkg replied:
It is explained right after, so no reason to repeat

pijuskri replied:
This can be resolved then

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
````diff
@@ -0,0 +1,998 @@
+	)
+```
+
+### `ORDER BY` clause
````
Perhaps explain the ordering rules per datatype

dahfjkg replied:
TBH, I think it is obvious without explanation. Data grids support sorting, so users should be familiar with how it works

pijuskri replied:
Can be resolved

pijuskri commented on content/en/docs/refguide/modeling/data/oql/oql-clauses.md:
```diff
@@ -0,0 +1,998 @@
+
+### `LIMIT` and `OFFSET` clauses
+
+With the `LIMIT` and `OFFSET` clauses, a portion of the result of a query can be returned. It is recommended to combine `LIMIT` and `OFFSET` clauses with `ORDER BY` because otherwise it can be impossible to predict which portion exactly weill be returned.
```
Perhaps a better phase would be "because the return the order of rows without any order by is inconsistent and can lead to unpredictable output"

pijuskri replied:
This can be resolved

pijuskri replied:
Can be resolved

MarkvanMents committed:
Move to new OQL section of Domain Model

MarkvanMents committed:
Proofread

MarkvanMents committed:
Proofread

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1000 @@
+Try your OQL example online with the [OQL Playground](https://service.mendixcloud.com/p/OQL) demo app. 
+
+## OQL Clauses
+
```
I think this is more appropriate for the OQL landing page. So we have the structure:

    OQL Landing Page (what is OQL where do you use it, how to use it - or links to pages on how to use it
        OQL Clauses
        OQL Expressions
        (other OQL docs?)

I've put a dummy OQL Landing page there for now - we can use some of what is in the existing OQL page, but I think we should ideally have a completely new page?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1000 @@
+
+### `SELECT` clause
+
+The `SELECT` clause specifies which entity attributes or other specified data must be retrieved. The `SELECT` clause consists of the term `SELECT` and one or more expressions. Each expression must be separated by a comma. Each expression defines a column in the result. Each expression can have an alias, which will be the name of the column in the result.
```
You say that you have `SELECT` plus expressions. However, you explain these expressions in this document, not in the **OQL Expressions** document. I think this is a bit confusing - is there another word we can use (sub-clauses, or definitions?) instead of **expressions** here?

Also, you say that _Each expression defines a column in the result._ However, the `*` expression defines multiple columns in the result, doesn't it?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1000 @@
+
+These basic clauses define which data needs to be retrieved and which entity or entities should be used as a source. Other clauses are not mandatory. They are used to specify limitations and other rules on the data being retrieved.
+
+Other clauses are not mandatory, but if used, they should be present in a fixed order:
```
```suggestion
These basic clauses define the data that needs to be retrieved and which entity or entities should be used as a source. Other clauses are not mandatory. They are used to specify limitations and other rules on the data being retrieved. If they are used, they should be present in the following order:
```

MarkvanMents reviewed (commented):
Part way through the review
Just noting here that I've moved the document to match the structure we discussed in Confluence. Will continue working on this tomorrow.

pijuskri committed:
Add literals, update COALESCE

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
````diff
@@ -0,0 +1,1000 @@
+SELECT Sales.Request/*
+FROM Sales.Customer
+JOIN Sales.Customer/Sales.Customer_Request/Sales.Request
+```
````
Using SQL for the language here has a weird effect on the markdown formatting as `/*` is not part of the SQL standard. We may have to not specify a language on these code blocks.
![image](https://github.com/user-attachments/assets/ad16a583-b347-4dac-b7b5-ed81190c3148)

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1000 @@
+
+
+{{% alert color="info" %}}
+In some scenarios, aliases in `FROM` are mandatory. Aliases are mandatory when defining a view entity.
```
Is this the only case of a mandatory use of an alias, or just an example?
If there are more, then can we define them all?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1000 @@
+
+`*` (asterisk) specifies that all attributes from all entities in the `FROM` clause should be returned. `entity_name/*` and `from_alias/*` specify that all attributes of the specified entity or expression of the `FROM` clause should be returned. `entity_name` can be optionally put in double quotes. If the entity name is a [reserved OQL word](/refguide/oql/#reserved-oql-words) (like `Order` or `Gr ...
+
+##### Examples
```
Do we need four examples. Can we just have two examples: the simplest `SELECT *` and the most complex?

dahfjkg replied:
I reduced the number of examples

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
````diff
@@ -0,0 +1,1000 @@
+```sql
+SELECT *
+FROM Sales.Customer
+JOIN Sales.Customer/Sales.Customer_Request/Sales.Request
````
Should we have an example of JOIN before we have documented it? I think this would be better in the **JOIN** section.

dahfjkg replied:
I removed this example, but other ones are used to illustrate `TableName/*`

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1000 @@
+
+| CustomerCount |
+| ----------    |
+| 5             |
```
I think these should be in the expressions doc. We don't need them here.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1000 @@
+| ----------    |
+| 5             |
+
+Another possibility is subqueries. A subquery can refer to other tables and expressions in `FROM`.
```
We currently have a whole document on subqueries (https://docs.mendix.com/refguide/oql-subqueries/). I think we should just move that into the OQL section of the documentation and refer to that rather than having examples here?

MarkvanMents committed:
Proofread

MarkvanMents reviewed (commented):
Hi @dahfjkg and @pijuskri 
I have done some proofreading in these commits: https://github.com/mendix/docs/pull/8431/files/8786adfd64abf1412b116d6433dd8457a11e7712..239aebdb960b3c3a018fedc47c43b47f5021afb3.

I have also made various suggestions in some comments.

Can you have a look at what I've done so far and see if you think I'm on the right track. We can discuss it further in a call if we need to discuss my changes.

If any of your earlier discussions are resolved can you mark them as resolved so it is clear what is still to do?

MarkvanMents committed:
Move oql-expressions to domain-model/oql

MarkvanMents committed:
Proofread oql-expressions

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1022 @@
+
+It is possible to use more complex expressions in `SELECT`. That is explained in detail in [OQL Expressions](/refguide/oql-expressions). Here are some examples:
+
+It is possible to use a function of an attribute. For instance:
```
Not sure about this. I initially didn't like the word "function" but I see that SQL has a lot of functions and that we are using it here in a similar way.
But the function isn't a function of the attribute (unlike a method in OOP) so perhaps 
```suggestion
It is possible to apply a function to an attribute. For instance:
```

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
````diff
@@ -0,0 +1,1022 @@
+
+```sql
+SELECT Cust/LastName
+FROM Sales.Customer Cust
````
Would this be clearer if we included `AS`?
```suggestion
FROM Sales.Customer AS Cust
```

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
````diff
@@ -0,0 +1,1022 @@
+JOIN Cust/Sales.Customer_Request/Sales.Request Req
+```
+
+#### Selecting Distinct Values with `DISTINCT`
````
This section is being referenced as #Distinct by oql-expressions

pijuskri replied:
This can be resolved

pijuskri committed:
Improve wording, change example order

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
````diff
@@ -0,0 +1,1022 @@
+| 562949953421923   | Jane      | Moose    | 1688849860264231 | Moose        | 2      |
+
+It is possible to specify columns in `SELECT`. To avoid ambiguity in case opf duplicate attribute names, it is possible to specify entity name or alias. Respectively, `<module>.<entity>/<attribute>` or `<alias>/<attribute>`. It is possible to retrieve alll attributes of a particular entity using `<alias>/*`
+```sql
````
I think we will have to leave the syntax highlighting off this as the `*` messes up the formatting and it looks weird (https://meta.stackoverflow.com/questions/401610/code-block-containing-asterisks-displaying-in-italics).
So
```suggestion
```
```

MarkvanMents committed:
Proofread

MarkvanMents reviewed (commented):
Got a bit further, but still have only got as far as JOIN.
Sorry this is taking so long - need to do other things at the same time!

pijuskri committed:
Add reference tag to aggregates section

pijuskri committed:
Align definition of expressions with oql clauses doc

dahfjkg committed:
Resolve review comments

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/mendix/docs?pullRequest=8431) <br/>All committers have signed the CLA.

pijuskri committed:
Add section on system variables

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1130 @@
+
+`entity_path` specifies the entity to join and the path from an earlier defined entity in the `FROM` clause to this entity.
+
+The example path `Crm.Customer/Crm.Customer_Address/Crm.Address` defines an association path from the entity **Crm.Customer** to a new entity **Crm.Address**.
```
```suggestion
The example path `Crm.Customer/Crm.Customer_Address/Crm.Address` defines an association path from the entity **Crm.Customer** to an associated entity **Crm.Address**.
```

pijuskri committed:
Add function examples

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1130 @@
+| Rekall | 9        |
+| Rekall | 3        |
+| Veidt  | 24       |
+| Veidt  | 2        |
```
Why does this table not include the `City` column?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1130 @@
+| Rekall | Utrecht   | 4              |
+| Veidt  | Utrecht   | 5              |
+
+### Subqueries
```
Will this replace the current [OQL Subqueries page](https://docs.mendix.com/refguide/oql-subqueries/)?

That page has subquery in HAVING. Are we deliberately not documenting this because subqueries in GROUP BY are database dependent?

dahfjkg replied:
GROUP by is database dependent, but HAVING should work in every database. I'm pushing some examples for that.

MarkvanMents committed:
Proofread to Subqueries section.

MarkvanMents reviewed (commented):
Hi there,
I've nearly finished my review. Sorry it is taking so long, but I keep being called away to do other things.
I have a couple more questions which are related to parts of the text. See below:

dahfjkg committed:
Update content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md

Co-authored-by: Mark van Ments <35492184+MarkvanMents@users.noreply.github.com>

dahfjkg committed:
Better examples for GROUP BY and HAVING

pijuskri committed:
Add range, cast examples

pijuskri committed:
Move file to correct location

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/_index.md:
```diff
@@ -0,0 +1,23 @@
+
+* `SELECT LastName FROM Sales.Customer` – retrieves the names of all customers
+* `SELECT FirstName FROM Sales.Customer WHERE LastName = 'Jansen'` – retrieves the first name of all customers with name "Jansen"
+* `SELECT SUM(TotalAmount) FROM Sales."Order" WHERE IsPaid = true` – retrieves the sum of the total prices of all paid orders (`Order` needs to be wrapped in quotes, see the [Reserved Words](#reserved-oql-words) section below)
```
There is no `reserved-oql-words` section

pijuskri replied:
Can be resolved

dahfjkg committed:
Describe OQL v2 behavior changes

dahfjkg committed:
Update content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md

Co-authored-by: Mark van Ments <35492184+MarkvanMents@users.noreply.github.com>

dahfjkg committed:
review comments

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/_index.md:
```diff
@@ -21,3 +21,79 @@ OQL queries do not take security into account out-of-the-box. This means that yo
+
+Words with a specific purpose in OQL are reserved. If you use reserved words for entity, variable or attribute names in an OQL query, they must be wrapped in double quotes `" "`. For example, in the OQL query `SELECT AVG(TotalPrice) FROM Sales."Order" WHERE IsPaid = 1`, `Order` needs to be wrapped in quotes because it is a reserved word, as it can be used for `ORDER BY`.
+
+Here is a list of all OQL reserved words:
```
Consider my notes on reserved words.
https://mendix.atlassian.net/wiki/spaces/DS/pages/3107586121/Undocumented+features+reserved+words

Some of the keywords in the list are used as function names, but are not strictly reserved. They are only used with function call syntax and so will not cause issues when used as regular identifiers. Might be worth to mention in a separate "keywords" list.

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
````diff
@@ -0,0 +1,136 @@
+
+The following is **not allowed**:
+
+```sql
````
This is actually allowed, see test in `SubqueryInJoin:339`

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,136 @@
+
+### `ORDER BY` without `LIMIT` and `OFFSET` in View Entity is not allowed
+
+For view entities, having `ORDER BY` without `LIMIT` and/or `OFFSET` is not allowed even for the top level query. Similarly to the case of subqueries, view entitiy results are not accessed directly, but are often subject of further filtering and ordering. TRherefore, it would not make sense to have `ORDER BY` directly in the view entity definition unless it is used to select a slice of the result ...
```
Maybe can be more specific and say "view entitiy results are not accessed directly, but are used as subqueries in OQL or xpath queries"

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,136 @@
+weight: 100
+---
+
+## OQL v2 features
```
GIven its a breaking change, i think this is important to mention also
https://mendix.atlassian.net/browse/DAT-3753

dahfjkg committed:
More review comments and more OQL v2 changes

pijuskri committed:
Handle review comments

pijuskri committed:
Handle review comments

Distinct examples, Expression definition, use cases, aggregate syntax, example aliases

pijuskri commented:
Passed technical review

MarkvanMents committed:
Proofread and confirm review changes

MarkvanMents committed:
Spellcheck and Lint

MarkvanMents committed:
Proofread oql/_index

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1234 @@
+
+The `SELECT` clause specifies which entity attributes or other specified data must be retrieved. The clause returns all the values of objects which match the `SELECT` clause.
+
+The `SELECT` clause consists of the term `SELECT` and one or more column definitions. Each column definition must be separated by a comma. Each column definition defines a column or a set of columns in the result. Each single value column definition can have an alias, which will be the name of the column in the result.
```
I think we need to say somewhere that a "column" is the same as an "attribute"?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1234 @@
+`entity_name` can optionally be put in double quotes. If the entity name is a [reserved OQL word](/refguide/oql-clauses/#reserved-oql-words) (like `Order` or `Group`), double quotes are mandatory.
+
+{{% alert color="warning" %}}
+Specifying all attributes will also return attributes which are normally hidden in the Domain Model, such as the `ID` of each object.
```
Is this a warning or just info?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1234 @@
+The `ORDER BY` clause specifies the sort order used on columns returned in a `SELECT` statement. Multiple columns can be specified. Columns are ordered in the sequence of the items in the `ORDER BY` clause.
+
+{{% alert color="info" %}}
+This clause can include items that do not appear in the `SELECT` clause, except when `SELECT DISTINCT` is specified or when a `GROUP BY` clause exists. When `UNION` is used, the column names or aliases must be those specified in the `SELECT` clause of the first part of the query.
```
I think we should add a link to UNION documentation (not sure where that is).

dahfjkg replied:
More than that: we should add UNION documentation itself 😅

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1234 @@
+
+#### `LIMIT` clause
+
+`LIMIT` specifies how many rows must be returned.
```
Isn't it the maximum number of rows that are returned. It could be less than LIMIT

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1234 @@
+| -1            | 0             |
+
+{{% alert color="info" %}}
+If you use a subquery as an expression in `SELECT`, then such subquery should always return a single column, and the number of rows should be at most one. If the subquery returns more than one row or a number of columns different than one, that will lead to an exception during runtime.
```
Could we just say a single value? Or saying the same thing in a shorter way:
```suggestion
If you use a subquery as an expression in `SELECT`, then such subquery should always return a single row and column. If the subquery returns more than one row or column, that will lead to an exception during runtime.
```

MarkvanMents committed:
Proofread oql-clauses

MarkvanMents reviewed (commented):
I have reviewed the clauses and _index documents again. My changes are here: https://github.com/mendix/docs/pull/8431/files/7915620e4eee2a89c1fde7f26fb10609adfcd284..04aec8a20f5c22496066d15e04c18b5c43641930 and I have left a couple more questions/suggestions.

I will try to do oql-v2 tomorrow and also the expression syntax.

At some stage it would be good to combine everything into one PR - but we can discuss this later.

MarkvanMents committed:
Proofread OQL V2

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,224 @@
+---
+title: "OQL Version 2 Features"
+linktitle: "OQL V2 Features"
```
Is this really "features". I would say it is "changes" or "differences".
You could even call it something like "Switching to OQL Version 2"?

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,224 @@
+
+### `ORDER BY` in Subquery
+
+You must now have a `LIMIT` and `OFFSET` in subquery containing `ORDER BY`. Using `ORDER BY` in subquery makes sense only when it is combined with `LIMIT` and `OFFSET`. Without the limitations, database engines do not guarantee that the row order in the subquery will be preserved in the outer query.
```
Is this **LIMIT and OFFSET** or **LIMIT and/or OFFSET**. You say both in the document.

MarkvanMents reviewed (commented):
I've now proofread oql-v2 in this commit: 2954bd39a0f67f15694198b75779f57eb0ae2e58
I have a couple of comments which need clarification.
I will try to read the third PR, #8520 before the end of the week.

MarkvanMents committed:
Spelling and Markdown Review

MarkvanMents committed:
Proofread to 5.4

MarkvanMents committed:
Proofread to COALESCE

MarkvanMents committed:
Finish Proofread of oql-expression-syntax

dahfjkg committed:
Address review comments

dahfjkg committed:
Update content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md

Co-authored-by: Mark van Ments <35492184+MarkvanMents@users.noreply.github.com>

dahfjkg commented:
I see that I still need to:
1.  Document `UNION`
2. Explain more bugfixes in OQL v2

dahfjkg committed:
Address review comments for Expressions Syntax

dahfjkg committed:
Address more review comments for Expressions Syntax

dahfjkg committed:
Document data type validations of expressions in OQL v2

dahfjkg committed:
Add links to OQL v2 validations to corresponding functions

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
````diff
@@ -0,0 +1,253 @@
+
+Instead, you can use the long path in the `JOIN` pattern as follows:
+
+```sql
+SELECT COUNT(*)
+FROM Module.Person AS P
+JOIN Module.Person/Module.Person_City/Module.City AS C
+GROUP BY C/Name
+```
````
Will this work, or does COUNT(*) need an **AS**?

hugomendix replied:
It will need an `AS`.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,253 @@
+SELECT
+  Name,
+  (
+    SELECT COUNT(*)
+    FROM Module.Person
+    WHERE Module.Person/City = Module.City/Name
```
Will this work, or does COUNT(*) need an AS?

hugomendix replied:
Same here; requires `AS`, because otherwise `COUNT(*)` would be a nameless column.

pijuskri replied:
`AS` is not required here, the column name will be the alias of the subquery `PersonCount`, the internal subquery column name has no effect on the main queries column names.

hugomendix replied:
Thanks. I missed the fact that it was nested from the diff.

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1258 @@
+SELECT
+	Req/Number AS RequestNumber,
+	(
+		SELECT COUNT(*)
+		FROM Sales.Customer AS Cust
+		WHERE Cust/LastName = Req/CustomerName
```
Will this work, or does COUNT(*) need an AS?

hugomendix replied:
`AS` needed.

pijuskri replied:
`AS` is not required, same reason as other case with subquery in attribute

dahfjkg replied:
In OQL v2, AS should be required

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1258 @@
+GROUP BY City
+HAVING
+	SUM(Stock) <= (
+		SELECT COUNT(*)
+		FROM Sales.Location
```
Will this work, or does COUNT(*) need an AS?

hugomendix replied:
Same asnwer; `AS` needed.

pijuskri replied:
`AS` is not required, `HAVING` and `WHERE` are not affected by changes to mandatory column names. It only applies to the main `SELECT` query and the attributes/columns of it.

MarkvanMents reviewed (commented):
Reviewed for **COUNT (*)** as discussed on Slack.

pijuskri committed:
Correct examples with missing attribute aliases

pijuskri committed:
Add oql UNION clause documentation section

pijuskri committed:
Improve oql-v2 docs on alias requirements

Extra information for union clause

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -1027,6 +1027,128 @@ OFFSET 2
+The clause takes multiple select queries and combines their results into a single result set.
+The resulting set by default only includes distinct rows. The `ALL` keyword can be used to include all rows. Rows are considered distinct if the combination of the column values is distinct from all other rows. Comparison logic of values is the same as the `DISTINCT` keyword of a `SELECT` clause.
+
+All select queries must define the same number of columns in the same order and their datatypes should match. The resulting column names will those used in the very first `SELECT` clause.
```
There is probably a word missing in `The resulting column names will those used`

MarkvanMents replied:
```suggestion
All select queries must define the same number of columns in the same order and their datatypes should match. The resulting column names will be those used in the very first `SELECT` clause.
```

dahfjkg commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -1027,6 +1027,128 @@ OFFSET 2
 | Rekall | Utrecht   | 4              |
 | Veidt  | Utrecht   | 5              |
 
+## `UNION` Clause {#oql-union}
```
We should add an example of how `UNION` works with association columns

passalaqua commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:
```diff
@@ -0,0 +1,1157 @@
+
+`expression` specifies the expression to convert.
+
+##### data_type
```
We still support CAST to FLOAT (and FLOAT is still a valid keyword in OQL). We have only deprecated it as a valid result type for queries, but intermediate calculations should still work.

MarkvanMents replied:
Deprecated or removed entirely?
Is FLOAT deprecated (will it go in Mx 11, for example). I can't really see any point in keeping it if we have DECIMAL.

pijuskri replied:
Is there a use case for a new user of OQL to use `FLOAT` instead of `DECIMAL` ?

passalaqua replied:
> Deprecated or removed entirely?

As I said:
> We have only deprecated it as a valid result type for queries, but intermediate calculations should still work.

And indeed, we still have tests for it and it still works, AFAICS.

Either we document here or in the list of keywords, but completely omitting mentioning something that would otherwise break your query if you use as an identifier is not a good idea, IMO. I just wrote the comment here because `CAST` is the only place I remember still accepts `FLOAT`.

dahfjkg replied:
I pushed a mention of `FLOAT`

passalaqua replied:
Looks good to me!

pijuskri replied:
Wouldn't a better place for the mention of `FLOAT` be in the `CAST` section, given that is the only valid use case?

dahfjkg replied:
If we are to move away from Float, I'd not mention any use cases of it at all

pijuskri replied:
Sorry, I reread Augusto's comment. Given the reason to mention it is to explain why its a keyword, then the current placement is good. Nothing else needs to be added.

MarkvanMents committed:
Typo

MarkvanMents committed:
Merge branch 'development' into pr/dahfjkg/8431

MarkvanMents committed:
Merge branch 'dat/oql-clauses' of https://github.com/dahfjkg/docs into pr/dahfjkg/8431

MarkvanMents committed:
Proofread

pijuskri committed:
Add example about UNION of associations

MarkvanMents committed:
Delete old OQL Clauses

MarkvanMents committed:
Remove old OQL Expressions documentation

MarkvanMents committed:
rename data-sets document

MarkvanMents committed:
Move Datasets into Data/Entities section

MarkvanMents committed:
Rename Domain Model section to Data

MarkvanMents committed:
Move OQL in navigation.

MarkvanMents committed:
Reorganize reserved words

MarkvanMents committed:
Proofread

dahfjkg committed:
Mention that FLOAT is not to be used in OQL

sbont commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,278 @@
+
+## Introduction
+
+In Studio Pro version 10.17, we introduced OQL v2. You will have to switch to OQL v2 in order to enable View entities. See the [View entities](/refguide/app-settings/#enable-view-entities) app setting for more information.
```
Note that we've renamed this setting to "OQL version 2", and the anchor will be #oql-version-2.
See also our branch for View Entities at https://github.com/mendix/docs/compare/development...sbont:docs:dux-1-view-entities
(Also, I assume you might want to update the Studio Pro release version...)

dahfjkg committed:
More details on system variables in OQL

dahfjkg committed:
Change link to setting that enables OQL v2

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:
````diff
@@ -58,14 +58,16 @@ There is no direct support for `DATETIME` literals. For functions that take `DAT
 '[%SystemVariable%]'
 ```
 
 These variables can be used the same way as other expressions.
````
To be more explicit, we should mention that system variables like `%CurrentUser%` are resolved to id's of objects with corresponding type (`LONG` ?). Same with date system variables, are they just strings? How do intervals work (like `%DayLength%`)?

dahfjkg replied:
Pushed an update

dahfjkg committed:
Add more information on tokens in OQL

MarkvanMents committed:
Proofread

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:
```diff
@@ -0,0 +1,1218 @@
+| 's*'   | 'my_string'     | `STRING`             | String literal                                |
+| d+     | 5               | `INTEGER` and `LONG` | Natural number literal                        |
+| d+.d+  | 5.3             | `DECIMAL`            | Real number literal                           |
+|        | `NULL`          | N/A                  | NULL literal to represent non-existent values. OQL does not support the keyword `empty`. In XPath, it has identical purpose as `NULL`. |
```
@pijuskri 
Is it a bit confusing to talk about XPath here? Or are you saying that in XPath you can't use `NULL` and have to use `empty` and in OQL you can't use `empty` and have to use `NULL`?

In that case I think it would be better to say:
```suggestion
|        | `NULL`          | N/A                  | NULL literal to represent non-existent values. OQL uses `NULL` where the equivalent XPath expression would use `empty`. |
```

dahfjkg replied:
The behavior is that XPath supports both NULL and empty, and OQL supports only NULL, so your proposed change is not entirely correct.

The idea here was to highlight similarities and differences to `XPath Keywords and System Variables` page. We already received some feedback from customers participating in beta testing who are familiar with XPath and who expected some things to work the same way in OQL.

Now that I think of it, system variables have lots of similarities between XPath and OQL and are definitely worth being described from that perspective, but Keywords in XPath are quite different from Literals in OQL, so It's also fine to omit mentioning XPath here.

MarkvanMents replied:
@dahfjkg 
Thanks for getting back to me, Pavel.
I wonder if it is useful to have the comparison somewhere. Would it be worth having an additional page which describes the differences (and similarities) between XPath and OQL? 

I'm happy to put something about XPath here (or get rid of it totally), but if we had it in another place that might be better - and resolve the issues that customers have raised?

dahfjkg replied:
The other issues are being addressed by implementing new features, and aside from those and the ones in XPath Keywords and System Variables, not that many points come to mind at the moment. Maybe we can consider that for the future, but I don't think that there is enough reason to work on that now. I'd say let's keep it mentioned here.

MarkvanMents replied:
@dahfjkg - so if we keep it, and my description above is wrong, how about:
```suggestion
|        | `NULL`          | N/A                  | NULL literal to represent non-existent values. OQL only uses `NULL` where the equivalent XPath expression can also use `empty`. |
```
?

dahfjkg committed:
Update content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md

Co-authored-by: Mark van Ments <35492184+MarkvanMents@users.noreply.github.com>

MarkvanMents reviewed (approved):
Thanks for all the hard work on this.
I will approve now, as I have no outstanding questions. But please let me know if you want to add anything else, or if this fails to get into 10.19.😀

pijuskri commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,278 @@
+* Long
+* Integer
+
+If any side of the operation is of a non-numeric type, no casting is performed, and the result is handled by the , as in OQL v1. See [Expression syntax](/refguide/oql-expression-syntax/#type-coercion) for more information.
```
There seems to be a missing word after `the result is handled by the `. @dahfjkg was it supposed to be "database"?

pijuskri committed:
Add oql UNION with differing data types documentation

Add mention in oql-v2 docs

pijuskri commented:
Since this PR is ready to be merged, for new content I created a separate PR targeting the oql-clauses branch on Pavels repository.
https://github.com/dahfjkg/docs/pull/3/files

MarkvanMents committed:
Merge branch 'development' into pr/dahfjkg/8431

dahfjkg committed:
Fix typo

dahfjkg committed:
Merge pull request #3 from pijuskri/dat/oql-union

Add documentation for UNION with differing data types

dahfjkg committed:
Explain how many-to-X associations in FROM lead to multiple rows

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
````diff
@@ -0,0 +1,1491 @@
+FROM Sales.Customer
+```
+
+In the case when the association [multiplicity type](/refguide/association-properties/#multiplicity) is one-to-many or many-to-many, every attribute over association will result in multiple results per object. If there are multiple attributes over association in the `FROM` clause, the result will be a cartesian product of associated objects meaning that there will be a row for every combination o ...
````
```suggestion
In the case when the association [multiplicity type](/refguide/association-properties/#multiplicity) is one-to-many or many-to-many, every attribute over association may result in multiple results per object. If there are multiple attributes over association in the `FROM` clause, the result will be a cartesian product of associated objects meaning that there will be a row for every combination of associated objects. If you want to avoid that effect, consider rewriting the query using [`JOIN`](/refguide/oql-clauses/#join).
```

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md:
```diff
@@ -0,0 +1,1491 @@
+
+In the case when the association [multiplicity type](/refguide/association-properties/#multiplicity) is one-to-many or many-to-many, every attribute over association will result in multiple results per object. If there are multiple attributes over association in the `FROM` clause, the result will be a cartesian product of associated objects meaning that there will be a row for every combination o ...
+
+For example, when a `Sales.Customer` with `LastName` "Doe" has two associated objects in `Sales.Request` with `Number` 1 and 2, the following query will return 4 rows:
```
```suggestion
For example, when a `Sales.Customer` with `LastName` "Doe" has two associated  `Sales.Request` objects with a `Number` attribute with values 1 and 2, the following query will return 4 rows:
```

MarkvanMents reviewed (commented):
Looks like a good addition. Have made a couple of proposals for changing the text.

dahfjkg committed:
Update content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md

Co-authored-by: Mark van Ments <35492184+MarkvanMents@users.noreply.github.com>

dahfjkg committed:
Update content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md

Co-authored-by: Mark van Ments <35492184+MarkvanMents@users.noreply.github.com>

MarkvanMents committed:
Add links to view entities.

MarkvanMents committed:
Merge branch 'dat/oql-clauses' of https://github.com/dahfjkg/docs into pr/dahfjkg/8431

MarkvanMents committed:
Merge branch 'development' into pr/dahfjkg/8431

MarkvanMents commented on content/en/docs/refguide/modeling/domain-model/oql/oql-v2.md:
```diff
@@ -0,0 +1,294 @@
+
+## Introduction
+
+In Studio Pro version 10.19, we introduced OQL v2. It can be enabled by setting the [OQL version 2](/refguide/app-settings/#oql-version-2) app setting to *Yes*. You will have to switch to OQL v2 in order to enable [View entities](/refguide/view-entities/).
```
@dahfjkg and @pijuskri 
The link to app-settings doesn't work and I can't find any other PR where this is added.
Is there supposed to be an update to app-settings in this PR?

MarkvanMents replied:
OK - found the link in another PR!

MarkvanMents committed:
Fix broken links.

MarkvanMents committed:
Merge branch 'development' into pr/dahfjkg/8431

MarkvanMents committed:
Broken link

MarkvanMents committed:
Add anchor
